Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp

```diff
@@ -1309,9 +1321,50 @@ void NetworkStorageManager::setStorageSiteValidationEnabled(bool enabled)
     });
 }
 
+void NetworkStorageManager::addAllowedSitesForConnectionInternal(IPC::Connection::UniqueID connection, const Vector<WebCore::RegistrableDomain>& sites)
+{
+    assertIsCurrent(workQueue());
+
+    if (!m_allowedSitesForConnections)
+        return;
+
+    auto& allowedSites = m_allowedSitesForConnections->ensure(connection, []() {
+        return HashSet<WebCore::RegistrableDomain> { };
+    }).iterator->value;
+
+    for (auto& site : sites)
+        allowedSites.add(site);
+}
+
+void NetworkStorageManager::addAllowedSitesForConnection(IPC::Connection::UniqueID connection, const Vector<WebCore::RegistrableDomain>& sites)
+{
+    ASSERT(RunLoop::isMain());
+    ASSERT(!m_closed);
+
+    if (sites.isEmpty())
+        return;
+
+    m_queue->dispatch([weakThis = ThreadSafeWeakPtr { *this }, connection, sites = crossThreadCopy(sites)]() mutable {
+        if (RefPtr protectedThis = weakThis.get())
+            protectedThis->addAllowedSitesForConnectionInternal(connection, sites);
+    });
+}
+
+bool NetworkStorageManager::isSiteAllowedForConnection(IPC::Connection::UniqueID connection, const WebCore::RegistrableDomain& site)
+{
+    assertIsCurrent(workQueue());
+
+    if (!m_allowedSitesForConnections)
+        return true;
+
+    auto allowedSites = m_allowedSitesForConnections->get(connection);
```

Comment:
We should try and avoid the HashSet copy here. One way to do this would be to use `HashMap::find()` instead of `get()`